util/webhook/validator.go

```diff
@@ -136,3 +144,43 @@ func validationResponseFromStatus(allowed bool, status metav1.Status) admission.
 		},
 	}
 }
+
+// ValidateImmutable validates equality across two values,
+// and returns a meaningful error to indicate a changed value, a newly set value, or a newly unset value.
+func ValidateImmutable(path *field.Path, oldVal, newVal any) *field.Error {
+	var unsetVal any
+	if reflect.TypeOf(oldVal) != reflect.TypeOf(newVal) {
+		return field.Invalid(path, newVal, "unexpected error")
+	}
+	switch oldVal.(type) {
+	case *string:
+		var t *string
+		unsetVal = t
+	case int32:
+		var t int32
+		unsetVal = t
+	case string:
+		var t string
+		unsetVal = t
+	case *bool:
+		var t *bool
+		unsetVal = t
+	default:
+		unsetVal = nil
```

Comment:
For any type not explicitly listed here, I think every comparison of equality to `unsetVal` will be false because a nil slice (or any typed `nil`) is not equivalent to untyped `nil`. I think this example demonstrates this: https://go.dev/play/p/0UQbIuroQeh

Overall I don't think this is causing any major problems since the `reflect.DeepEqual` call is catching the type discrepancy, but it does hide the "can't set previously unset field"/"can't unset previously set field" context.

I tried modifying the AzureMachine webhook tests to print the exact error being returned and found that to be the case. Removing this definition of a string yields `field is immutable, unable to assign a value if it was already empty`:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d77c067f5542624a0de5d0b8c0288d0f7fbc74fa/api/v1beta1/azuremachine_webhook_test.go#L232
And removing this definition of a struct pointer yields only `field is immutable`:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d77c067f5542624a0de5d0b8c0288d0f7fbc74fa/api/v1beta1/azuremachine_webhook_test.go#L132-L134

Reply:
These observations match mine. For types that we don't explicitly have a case for, we're essentially wrapping `reflect.DeepEqual`. Strictly speaking this adds a bit of overhead when we change to this new func, with the benefit of using a common boilerplate pattern.

Reply:
I think that's reasonable to essentially default to `reflect.DeepEqual`. If we can boil this down into a comment that makes sense I think that would be worth adding.